From 3b61560835cbcb203c980d7c47f74ab0868a862f Mon Sep 17 00:00:00 2001
From: Zhu Ken <guigenz@codeaurora.org>
Date: Tue, 22 Aug 2017 13:41:46 -0700
Subject: [PATCH] map: add support BR mode and add option fmr

There are something different between BR and CE of MAP
and add fmr option for setting the fmr rules

Change-Id: Ic99058be51f47f3c30f369085082937d29713149
Signed-off-by: Zhu Ken <guigenz@codeaurora.org>
(cherry picked from commit 7162ea10391629cc791696514bbb12064a95b76f)
Signed-off-by: Ken Zhu <quic_guigenz@quicinc.com>
---
 package/network/ipv6/map/files/map.sh | 27 +++++++++++++++++++++------
 1 file changed, 21 insertions(+), 6 deletions(-)

diff --git a/package/network/ipv6/map/files/map.sh b/package/network/ipv6/map/files/map.sh
index 652c3065e3da..abd4f631feae 100755
--- a/package/network/ipv6/map/files/map.sh
+++ b/package/network/ipv6/map/files/map.sh
@@ -26,8 +26,8 @@ proto_map_setup() {
 	local link="map-$cfg"
 
 	local maptype type legacymap mtu ttl tunlink zone encaplimit
-	local rule ipaddr ip4prefixlen ip6prefix ip6prefixlen peeraddr ealen psidlen psid offset
-	json_get_vars maptype type legacymap mtu ttl tunlink zone encaplimit
+	local rule ipaddr ip4prefixlen ip6prefix ip6prefixlen peeraddr ealen psidlen psid offset mode fmr
+	json_get_vars maptype type legacymap mtu ttl tunlink zone fmr mode encaplimit
 	json_get_vars rule ipaddr ip4prefixlen ip6prefix ip6prefixlen peeraddr ealen psidlen psid offset
 
 	[ "$zone" = "-" ] && zone=""
@@ -52,6 +52,7 @@ proto_map_setup() {
 			rule="$rule,dmr=$peeraddr"
 		else
 			rule="$rule,br=$peeraddr"
+			[ $fmr = 1 ] && rule="$rule,fmr=1"
 		fi
 	fi
 
@@ -81,8 +82,14 @@ proto_map_setup() {
 		json_add_string mode ipip6
 		json_add_int mtu "${mtu:-1280}"
 		json_add_int ttl "${ttl:-64}"
-		json_add_string local $(eval "echo \$RULE_${k}_IPV6ADDR")
-		json_add_string remote $(eval "echo \$RULE_${k}_BR")
+		if [ "$mode" = br ]; then
+			json_add_string remote $(eval "echo \$RULE_${k}_IPV6ADDR")
+			json_add_string local $(eval "echo \$RULE_${k}_BR")
+		else
+			json_add_string local $(eval "echo \$RULE_${k}_IPV6ADDR")
+			json_add_string remote $(eval "echo \$RULE_${k}_BR")
+		fi
+
 		json_add_string link $(eval "echo \$RULE_${k}_PD6IFACE")
 		json_add_object "data"
 			[ -n "$encaplimit" ] && json_add_string encaplimit "$encaplimit"
@@ -132,6 +139,7 @@ proto_map_setup() {
 	[ -n "$zone" ] && json_add_string zone "$zone"
 
 	json_add_array firewall
+	if [ "$mode" != br ]; then
 	  if [ -z "$(eval "echo \$RULE_${k}_PORTSETS")" ]; then
 	    json_add_object ""
 	      json_add_string type nat
@@ -139,7 +147,7 @@ proto_map_setup() {
 	      json_add_string family inet
 	      json_add_string snat_ip $(eval "echo \$RULE_${k}_IPV4ADDR")
 	    json_close_object
-	  else
+	   else
 	    for portset in $(eval "echo \$RULE_${k}_PORTSETS"); do
               for proto in icmp tcp udp; do
 	        json_add_object ""
@@ -153,6 +161,7 @@ proto_map_setup() {
 	        json_close_object
               done
 	    done
+	   fi
 	  fi
 	  if [ "$maptype" = "map-t" ]; then
 		[ -z "$zone" ] && zone=$(fw3 -q network $iface 2>/dev/null)
@@ -192,7 +201,11 @@ proto_map_setup() {
 		json_add_string ifname "@$(eval "echo \$RULE_${k}_PD6IFACE")"
 		json_add_string proto "static"
 		json_add_array ip6addr
-		json_add_string "" "$(eval "echo \$RULE_${k}_IPV6ADDR")"
+		if [ "$mode" = br ]; then
+			json_add_string "" "$(eval "echo \$RULE_${k}_BR")"
+		else
+			json_add_string "" "$(eval "echo \$RULE_${k}_IPV6ADDR")"
+		fi
 		json_close_array
 		json_close_object
 		ubus call network add_dynamic "$(json_dump)"
@@ -238,6 +251,8 @@ proto_map_init_config() {
 	proto_config_add_int "ttl"
 	proto_config_add_string "zone"
 	proto_config_add_string "encaplimit"
+	proto_config_add_string "mode"
+	proto_config_add_int "fmr"
 }
 
 [ -n "$INCLUDE_ONLY" ] || {
-- 
2.34.1

